src/engine/TPA.cc

```diff
@@ -1379,98 +1413,115 @@ void TransitionSystemNetworkManager::initNetwork() {
+    auto graphEnds = getGraphEnds();
+    bool safe;
+    for(auto init: graphStarts){
+        auto current = init;
+        activePath.clear();
+        activePath.push(getIncomingEdge(current, 0));
+        safe = false;
+        while (!safe) {
+            if (getNode(current).blocked_children == getNode(current).children.size()) {
                 getNode(current).trulyReached = PTRef_Undef;
-//                std::cout << "Edge cannot propagate forward from: " << logic.pp(edgeExplanation) << std::endl;
-                getNode(current).solver->updateQueryStates(logic.mkNot(edgeExplanation));
-                continue; // Repeat the query for the same TS with stronger query
-            }
-        } else { // TS cannot propagate forward
-            if (current == getChainStart()) {
-                return {VerificationAnswer::SAFE, ValidityWitness{}};
+                getNode(current).blocked_children = 0;
+                updateRestrictions(current);
             }
-            assert(getNode(current).trulySafe != PTRef_Undef);
-//            std::cout << "Query not reachable in " << current.id << ". These states are safe: " << logic.pp(explanation) << std::endl;
-            getNode(current).trulySafe = logic.mkOr(getNode(current).trulySafe, explanation);
-            EId previousEdge = getIncomingEdge(current);
-            auto & previousNode = getNode(graph.getSource(previousEdge));
-            assert(previousNode.trulyReached != PTRef_Undef);
-            PTRef updatedConditions = logic.mkNot(getNode(current).trulySafe);
-            auto [edgeRes, edgeExplanation] = queryEdge(previousEdge, previousNode.trulyReached, updatedConditions);
-            if (reachable(edgeRes)) { // New reached, not refuted yet, states
-//                std::cout << "Found new potential initial states: " << logic.pp(edgeExplanation) << std::endl;
-                getNode(current).solver->resetInitialStates(edgeExplanation);
-                continue; // Repeat the query for the same TS with new initial states
-            } else { // Cannot continue from currently computed truly reached states
-                previousNode.trulyReached = PTRef_Undef;
-//                std::cout << "Strenghtening source TS with: " << logic.pp(edgeExplanation) << std::endl;
-                previousNode.solver->updateQueryStates(logic.mkNot(edgeExplanation));
-                current = graph.getSource(previousEdge);
-                continue; // Blocked states have been propagated to the previous node, repeat the query there.
+            auto [res,explanation] = queryTransitionSystem(networkMap.at(current));
+            if (reachable(res)) {
+                getNode(current).trulyReached = explanation;
+                for(auto end: graphEnds){
+                    if (current == end) {
+                        return {VerificationAnswer::UNSAFE, computeInvalidityWitness()};
+                    }
+                }
+                EId nextEdge;
```

Comment:
This is used only inside the loop, no? No need to declare it outside the loop.

Reply:
I think it saves some time for the performance, because it doesn't need to remove and create it each iteration